src/irmin/contents.ml

```diff
@@ -121,7 +109,27 @@ module Json_value = struct
     |~ case1 "array" (list ty) (fun arr -> `A arr)
     |> sealv)
 
-  let t = Type.like' ~cli:(pp, of_string) t
+  let rec equal a b =
+      match a, b with
+      | `Null, `Null -> true
+      | `Bool a, `Bool b -> Type.(equal bool) a b
+      | `String a, `String b -> String.equal a b
+      | `Float a, `Float b -> Type.(equal float) a b
+      | `A a, `A b ->
+          (try
+            List.for_all2 (fun a' b' ->
+              equal a' b') a b
+          with Invalid_argument _ -> false)
+      | `O a, `O b ->
+          (try
+            List.for_all2 (fun (k, v) (k', v') ->
+              equal (List.assoc k b) v && equal (List.assoc k' a) v') a b
```

Comment:
The `List.assoc` looks a bit strange. Maybe:

```ocaml
(try List.for_all2 (fun (k, v) (k', v') -> k=k' && equal v v') (List.sort compare_fst a) (List.sort compare_snd b) 
with Invalid_argument _ -> false)
```
e.g. sort the lists first, to avoid a quadratic complexity with `List.assoc`

Reply:
Thanks, that is much nicer!